Tyr OSRM serializer: create bannerInstructions in step

eikes opened the pull request:
# Issue

There is a proprietary extension of the OSRM routing format used in the Mabox Navi SDK and MapLibre Navi SDK, both in iOS and Android. These SDKs only work when the `bannerInstructions` attribute is present in the `steps`. This PR aims to optionally add the `bannerInstructions` behind a flag in the request.

## Tasklist

 - [x] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [x] Update the docs with any new request parameters or changes to behavior described
 - [x] Update the [changelog](CHANGELOG.md)
 - [ ] If you made changes to the lua files, update the [taginfo](taginfo.json) too.

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

kevinkreiser commented:
excellent, ive been meaning to add this for years to make the output more compatible with the mapbox apis again but never had the time.

ianthetechie commented:
Ooh nice. I see this is currently a draft PR. What is still needed to get this into a mergeable state @eikes? Anything we could do to help?

ianthetechie commented:
Bump @eikes?

eis-ioki commented:
@ianthetechie @kevinkreiser any idea why these CI builds are failing?

eikes commented:
@nilsnolde running `./scripts/format.sh` starts an endless loop of downloading and uncompressing `clang++/11.0.0.tar.gz`, `llvm/11.0.0.tar.gz`, `clang++/9.0.1.tar.gz` and `llvm/9.0.1.tar.gz`

nilsnolde commented:
Just wanna leave a quick thank you, for stepping down the rabbit hole :smile:

eikes commented:
Here is what I can say about the state of this PR and the banner instructions which it creates: 

1. The JSON structure of the `bannerInstructions` attribute matches what the navigation SDKs expect.
2. The idea of each banner component is implemented correctly. It shows the information about and the distance to the upcoming maneuver:
    1. The `primary` banner includes the streeet name / ref and the type of the maneuver
    2. The `secondary` banner shows the destination if available
    3. The `sub` banner shows the lane information if available
3. There is still room for improvement:
    1. I haven't implemented highway shields yet. The mapbox SDK can show highway shields using the `mapbox_shield` property (why this is not called highway_shields is unclear to me), and they are really nice and helpful, but they need to take the country and highway type into account. I think it's something that can be implemented in a subsequent PR.
    2. Lane instructions are currently only shown towards the end of the step. Maybe they can be helpful along the way of a longer step though. I am not sure if the navigation SDKs even support more than two elements in the `bannerInstructions` array though.
4. Verification: The banner instructions work with a recent version of the Mapbox android navigation SDK, I haven't tested it with the MapLibre navigation SDK yet, but I expect it to work with it as well. I will do that too.
5. It needs more verification, given that people could use this in a real navigation system. I will also do more manual inspection of various types of route descriptions and their maneuvers, such as roundabouts, on- and off-ramps and / or streets without names. I am planning on creating a tool to visually inspect the generated routes with the banners.

Other than that, I am quite happy with the progress so far and the positive feedback I have received here. Much appreciated, thank you!

eis-ioki commented:
Is anyone willing to give me some feedback on this PR?

kevinkreiser commented:
@eikes yes of course! i will be happy to review it as soon as possible. as you have seen we have a lot of things going on at once and there are only really 2 of us with full time jobs so its hard to really pick which thing to review when, very sorry about that. anyway we'll have a look soon. i did read over your last comment from last week and was wondering what you were going to say about the highway shields :smile: you're choice in forgoing them for the time being makes sense to me as they are a permutation nightmare (from recollection). as for the rest of it i'll look at it maybe this evening more likely this weekend

kevinkreiser commented on docs/docs/api/turn-by-turn/api-reference.md:
```diff
@@ -311,7 +319,6 @@ Directions options should be specified at the top level of the JSON object.
 | `exclude_locations` |  A set of locations to exclude or avoid within a route can be specified using a JSON array of avoid_locations. The avoid_locations have the same format as the locations list. At a minimum each avoid location must include latitude and longitude. The avoid_locations are mapped to the closest road or roads and these roads are excluded from the route path computation.|
 | `exclude_polygons` |  One or multiple exterior rings of polygons in the form of nested JSON arrays, e.g. `[[[lon1, lat1], [lon2,lat2]],[[lon1,lat1],[lon2,lat2]]]`. Roads intersecting these rings will be avoided during path finding. If you only need to avoid a few specific roads, it's **much** more efficient to use `exclude_locations`. Valhalla will close open rings (i.e. copy the first coording ...
 | `date_time` | This is the local date and time at the location.<ul><li>`type`<ul><li>0 - Current departure time.</li><li>1 - Specified departure time</li><li>2 - Specified arrival time. Not yet implemented for multimodal costing method.</li></li>3 - Invariant specified time. Time does not vary over the course of the path. Not implemented for multimodal or bike share routing</li></ul></li><li>`va ...
-| `out_format` | Output format. If no `out_format` is specified, JSON is returned. Future work includes PBF (protocol buffer) support. |
```
:pray:

kevinkreiser commented on docs/docs/api/turn-by-turn/api-reference.md:
```diff
@@ -268,8 +268,16 @@ Directions options should be specified at the top level of the JSON object.
 | `units` | Distance units for output. Allowable unit types are miles (or mi) and kilometers (or km). If no unit type is specified, the units default to kilometers. |
 | `language` | The language of the narration instructions based on the [IETF BCP 47](https://tools.ietf.org/html/bcp47) language tag string. If no language is specified or the specified language is unsupported, United States-based English (en-US) is used. [Currently supported language list](#supported-language-tags) |
 | `directions_type` |  An enum with 3 values. <ul><li>`none` indicating no maneuvers or instructions should be returned.</li><li>`maneuvers` indicating that only maneuvers be returned.</li><li>`instructions` indicating that maneuvers with instructions should be returned (this is the default if not specified).</li></ul> |
+| `format` | Three options are available: <ul><li>`json` is default valhalla routing directions JSON format</li><li>`gpx` returns the route as a GPX (GPS exchange format) XML track</li><li>`osrm` creates a OSRM compatible route directions JSON</li><li>`pbf` formats the result using protocol buffers</li></ul> |
```
its 4 right, though i seriously think we should just delete gpx

```suggestion
| `format` | Four options are available: <ul><li>`json` is default valhalla routing directions JSON format</li><li>`gpx` returns the route as a GPX (GPS exchange format) XML track</li><li>`osrm` creates a OSRM compatible route directions JSON</li><li>`pbf` formats the result using protocol buffers</li></ul> |
```

kevinkreiser commented:
should add voice_instructions after we get this reviewed and merged to complete the package :wink:

kevinkreiser commented on src/tyr/route_serializer_osrm.cc:
```diff
@@ -429,6 +429,54 @@ struct IntersectionEdges {
 
+// Process 'indications' array - add indications from left to right
+json::ArrayPtr lane_indications(std::unique_ptr<valhalla::odin::EnhancedTripLeg_Edge>& prev_edge,
+                                const valhalla::TurnLane* turn_lane) {
```
nit: why pass this by pointer? indeed why pass it at all? in the end you just pull the mask off of it, perhaps its best to just pass the mask to this function by value eg:
```suggestion
                                uint16_t mask) {
```

eikes replied:
Fair point! I made it even simpler by changing the signature to:

```CPP
lane_indications(const bool drive_on_right, const uint16_t mask)
```

kevinkreiser commented:
@eikes i started looking at this this morning. it would be vastly easier to review if you had a chance to go through the diff and make some comments about the changes and what they are doing/intended to do. i'll keep plugging along but some of the refactoring is hard to make sense of without some guiding commentary from you.

kevinkreiser commented on src/tyr/route_serializer_osrm.cc:
```diff
@@ -1197,11 +1170,187 @@ json::MapPtr osrm_maneuver(const valhalla::DirectionsLeg::Maneuver& maneuver,
+    osrm_man->emplace("modifier", modifier);
+  }
+  // Roundabout count
+  if (maneuver.type() == DirectionsLeg_Maneuver_Type_kRoundaboutEnter &&
```
ok so this gets pulled out of the guts of the maneuver function to the top level

eikes replied:
This was previously nested within the `osrm_maneuver` part where the `maneuver_type` was determined. `maneuver_type` had to become its own function, so it could be reused in the `banner_instructions` though. This was the only part that affected the `osrm_man`, so it had to be a part of the `osrm_maneuver` function.

kevinkreiser commented on src/tyr/route_serializer_osrm.cc:
```diff
@@ -1442,11 +1592,27 @@ json::ArrayPtr serialize_legs(const google::protobuf::RepeatedPtrField<valhalla:
         step->emplace("rotary_name", maneuver.street_name(0).value());
       }
 
+      // Get incoming and outgoing bearing. For the incoming heading, use the
+      // prior edge from the TripLeg. Compute turn modifier. TODO - reconcile
+      // turn degrees between Valhalla and OSRM
+      uint32_t idx = maneuver.begin_path_index();
+      uint32_t in_brg = (idx > 0) ? etp.GetPrevEdge(idx)->end_heading() : 0;
+      uint32_t out_brg = maneuver.begin_heading();
```
ok this got pulled out because the function we call now doesnt have the path so it cant look at the previous edge. kind of a mess but im not sure how to simplify it. i understand you wanted to not have one large `osrm_maneuver` method but it does kind of spread logic around. no real clean way to do it, either one large method or smaller ones but with state spread around between them :frowning_face:

eikes replied:
Yes. And also: both the `turn_modifier` and `osrm_maneuver` function utilise the bearing variables `in_brg` and `out_brg`. The `turn_modifier` was pulled out because it is being used in the `maneuver_type` in the `banner_instructions` function itself (along with the `modifier` from the `turn_modifier` call). 

I did all of that in order to **not** run the same code again when creating the `banner_instructions`.

kevinkreiser commented on proto/options.proto:
```diff
@@ -482,4 +482,7 @@ message Options {
     uint32 matrix_locations = 54;                                  // a one to many or many to one time distance matrix. Does not affect
   }                                                                // sources_to_targets when either sources or targets has more than 1 location
                                                                    // or when CostMatrix is the selected matrix mode.
+  oneof has_banner_instructions {
+    bool banner_instructions = 55;                                 // Whether to return bannerInstructions in the OSRM serializer response
+  }
```
unset defaulting to `false` is ok in this case, therefore we dont need `oneof`
```suggestion
  bool banner_instructions = 55;                                 // Whether to return 
```

kevinkreiser commented on test/gurka/test_osrm_serializer.cc:
```diff
@@ -515,3 +515,196 @@ TEST(Standalone, HeadingNumberAutoRoute) {
   // clang-format on
 }
+
+TEST(Standalone, BannerInstructions) {
```
ausführlich, respekt!

eikes replied:
I realised later on that I could have used `assert_json_equality(serialized_json, expected_json)` instead, but this works too...

eikes commented:
> @eikes i started looking at this this morning. it would be vastly easier to review if you had a chance to go through the diff and make some comments about the changes and what they are doing/intended to do. i'll keep plugging along but some of the refactoring is hard to make sense of without some guiding commentary from you.

@kevinkreiser I am not sure if it really makes sense to comment the individual lines, so here is what I did and why:

In order to create the object for the `bannerInstructions` some data that is already present in other places in the JSON is repeated. To avoid running the same code multiple times as much as possible, the code which determines the `turn_modifier` and the `maneuver_type` is moved into its own method and the result is stored and reused. The same goes for the `lane_indication` method from the `intersections` function which is reused to create the lane directions in the `sub` banner, this code does run twice, but I don't want to mix the intersections code with the `sub` code. (The only option would be to have an intermediate representation of the lanes, but I don't think it slows things down anyways.)

What the `banner_instructions` are concerned, I think it's mostly self-explanatory, except for maybe the `sub_banner_instruction`, in this method I go back from the end of the step through all the edges which have `turn_lanes` until one is found which doesn't. These lanes are then put into the `sub` banner. I still need to investigate whether or not it is possible to have more than two `bannerInstructions` per step, and if so we could provide separate banner instruction along the step if there are multiple lanes involved. The heuristic currently is, that if a step is longer than 400 meters, the sub banner is shown during the last 400 meters, otherwise it becomes a part of the first banner instruction. The construction of the additional banner instruction happens in the `banner_instructions` method.

kevinkreiser commented:
@eikes would you mind resolving the conflicts in the changelog. if you are happy with your progress i think we can merge early this week. i had a look over it and i think even if we find things that arent perfect with it its a great start!

eikes commented:
@kevinkreiser I have rebased this branch on to the latest master branch and there are no conflicts now.

eikes commented:
@kevinkreiser I am not quite happy with the roundabouts directions just yet. I would like to fix those before you merge.

kevinkreiser commented:
@eikes kein problem, sag bescheid

eikes commented:
@kevinkreiser I have added my changes for the roundabouts: The `degree` is supposed to represent how far along the circle of the roundabout you are supposed to drive, this can be used to draw an appropriate roundabout icon. If you go straight through a roundabout, the degree is 180, if you go from south to east (driving on the right side), the `degree` would be 90, and so forth. This is not information I was able to pull in from somewhere else, so I had to come up with a method to approximate it myself.

Right now, the degree is calculated by comparing the `end_heading` of the last edge before the roundabout with the `begin_heading` of the first edge after the roundabout. 

In real world scenarios, the `degree` is still not quite what I want though, as these headings are often quite shallow angles compared to the circle, as the incoming street eases into the roundabout and the exit eases out.

Ideally I would like to calculate the degree not based on these headings, but rather by looking at where in the roundabout circle the `roundabout enter` maneuver is and where in the roundabout circle the `roundabout exit` is. If I had any of the following information, I could do a better calculation:

1. The circumference of the entire roundabout
2. The LL-location of the center of the roundabout
3. The overall heading of the inbound and outbound streets. I realise this is a bit vague and can be hard to put into code.

My preferred method would be to simply compare the length of the trip maneuver on the rotary with the rotary circumference, but I don't know if I can inspect the graph like that when the route is serialised, so I'll add it to my list of things that can be improved and I can do it when I understand more of how Valhalla works :D (Maybe this information needs to be stored in the edges during the graph parsing phase?)

Other than that it's a good start and if you feel like this is ready to be merged, then please do. I'll get back to this after my summer vacation. If you have any changes you'd like to see please let me know, it might me a couple of weeks until I can address them though.

kevinkreiser commented:
@eikes i believe @dgearhart and another contributor messed with this for a long time back in the day... let me see... maybe this one: #1559. at this point it doesnt have any useful information as its a major version behind but i wanted to point it out because its a difficult problem when faced with "real life" data so dont feel bad about the approximate solution

eikes commented:
@kevinkreiser Agreed, this should happen in a separate PR. I'll figure out how to add it to the data processing stage, that makes the most sense.

eikes commented:
@kevinkreiser Will this get merged? Should I pull out the _optional_ degrees into a separate PR entirely, going back to the state it was at July 17th?

kevinkreiser commented:
no it's fine as it is i just wasn't aware you were ready for final review

eikes commented:
Yes, please do review this PR 😄

kevinkreiser commented on proto/options.proto:
```diff
@@ -482,4 +482,5 @@ message Options {
     uint32 matrix_locations = 54;                                  // a one to many or many to one time distance matrix. Does not affect
   }                                                                // sources_to_targets when either sources or targets has more than 1 location
                                                                    // or when CostMatrix is the selected matrix mode.
+  bool banner_instructions = 55;                                 // Whether to return 
```
whitespace nit:
```suggestion
  bool banner_instructions = 55;                                  // Whether to return 
```

kevinkreiser commented on CHANGELOG.md:
```diff
@@ -21,6 +21,7 @@
    * CHANGED: refactor landmark database interface to use a pimpl [#4202](https://github.com/valhalla/valhalla/pull/4202)
    * ADDED: support for `:forward` and `:backward` for `motor_vehicle`, `vehicle`, `foot` and `bicycle` tag prefixes [#4204](https://github.com/valhalla/valhalla/pull/4204)
    * ADDED: add `valhalla_build_landmarks` to parse POIs from osm pbfs and store them as landmarks in the landmark sqlite database [#4201](https://github.com/valhalla/valhalla/pull/4201)
+   * ADDED: Support for `bannerInstructions` attribute in OSRM serializer via `banner_instructions` request parameter [#4093](https://github.com/valhalla/valhalla/pull/4093)
```
can we move this to the bottom?

kevinkreiser reviewed (dismissed):
just some tiny nits and we can merge this

eis-ioki commented:
I implemented the changes ([9210c7d](https://github.com/valhalla/valhalla/pull/4093/commits/9210c7d96d0b30d2618178a35cabe323a21453bc)) and rebased the whole branch on the master branch once more. This somehow invalidated your review, but I addressed the change requests.

eikes commented:
I implemented the changes ([9210c7d](https://github.com/valhalla/valhalla/pull/4093/commits/9210c7d96d0b30d2618178a35cabe323a21453bc)) and rebased the whole branch on the master branch once more. This somehow invalidated your review, but I addressed the change requests.

eikes commented:
Sweet! Four months in the making! I am very happy to see this merged. Next up: `voiceInstructions` !
